### PR TITLE
[Protobuf] add a way to switch between RLP and Protobuf for substateDB decoding

### PR DESCRIPTION
**Pull request by rpl-ffl**
Squashed from #86 

Add a builder function `SetDecoder(encoding string)` to `substateDB` and `SubstateDB` interface so that a decoder type can be selected.
A substateDB can decode using `decodeRlp` or `decodeProtobuf`. It defaults to `decodeRlp` to preserve original functionality.

Example usage:
```
db := (&substateDB{...})          // initializing db as normal, this defaults to RLP
         .SetDecoder(<encoding>)  // at end of init (or before decoding) to switch between RLP/Protobuf
```

## Type of change
- [ ] New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: db/substate_decoder.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+			return decodeRlp(bytes, lookup, block, tx)
+		}
+	}
+}
+
+// decodeRlp decodes into substate the provided rlp-encoded bytecode
+func decodeRlp(bytes []byte, lookup codeLookup, block uint64, tx int) (*substate.Substate, error) {
+	rlpSubstate, err := rlp.Decode(bytes)
+	if err != nil {
+		return nil, fmt.Errorf("cannot decode data into rlp block: %v, tx %v; %w", block, tx, err)
```
Comment (cabrador) [inline db/substate_decoder.go:59]:
```suggestion
		return nil, fmt.Errorf("cannot decode substate data from rlp block: %v, tx %v; %w", block, tx, err)
```

File: db/substate_decoder.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+	}
+
+	return rlpSubstate.ToSubstate(lookup, block, tx)
+}
+
+// decodeProtobuf decodes into substate the provided protobuf-encoded bytecode
+func decodeProtobuf(bytes []byte, lookup codeLookup, block uint64, tx int) (*substate.Substate, error) {
+	pbSubstate := &pb.Substate{}
+	if err := proto.Unmarshal(bytes, pbSubstate); err != nil {
+		return nil, fmt.Errorf("cannot decode data into protobuf block: %v, tx %v; %w", block, tx, err)
```
Comment (cabrador) [inline db/substate_decoder.go:69]:
```suggestion
		return nil, fmt.Errorf("cannot decode substate data from protobuf block: %v, tx %v; %w", block, tx, err)
```

File: db/substate_db.go [hunk 3]
```diff
@@ -140,14 +137,9 @@ func (db *substateDB) GetBlockSubstates(block uint64) (map[int]*substate.Substat
...
-		pbSubstate := &pb.Substate{}
-		if err := proto.Unmarshal(value, pbSubstate); err != nil {
-			return nil, err
-		}
-
-		sbstt, err := pbSubstate.Decode(db.GetCode, block, tx)
+		sbstt, err := db.DecodeSubstate(value, block, tx)
 		if err != nil {
-			return nil, err
+			return nil, fmt.Errorf("Error decoding block %d, tx %d; %w", block, tx, err)
```
Comment (cabrador) [inline db/substate_db.go:142]:
```suggestion
			return nil, fmt.Errorf("failed to decode substate, block: %d, tx: %d; %w", block, tx, err)
```

**Review by cabrador — CHANGES_REQUESTED** [review]
Please add unit tests.

**Comment by rpl-ffl** [comment]
Deprecated in favor of https://github.com/Fantom-foundation/Substate/pull/88, as discussed.

**Pull request closed without merging**